keeper/src/main/scala/zio/keeper/transport/udp.scala

```diff
@@ -0,0 +1,137 @@
+  ): ZIO[Clock with Logging[String], Nothing, Transport] = {
+    val connectionTimeout_ = connectionTimeout
+    val requestTimeout_ = requestTimeout
+    val chunkSize_ = chunkSize
+    ZIO.environment[Clock with Logging[String]].map { env =>
+      new Live with Clock {
+        override val connectionTimeout: Duration = connectionTimeout_
+        override val requestTimeout: Duration = requestTimeout_
+        override val logger: Logging.Service[Any, String] = env.logging
+        override val clock: Clock.Service[Any] = env.clock
+        override val chunkSize: Int = chunkSize_
+      }
+    }
+  }
+
+  trait Live extends Transport { self: Clock =>
+    val connectionTimeout: Duration
+    val requestTimeout: Duration
+    val chunkSize: Int
+
+    val logger: Logging.Service[Any, String]
+
+    val transport = new Transport.Service[Any] {
+
+      override def connect(to: SocketAddress): ZManaged[Any, TransportError, ChannelOut] =
+        (for {
+          datagramChannelAndClose  <- DatagramChannel().withEarlyRelease.mapError(ExceptionWrapper)
+          (close, datagramChannel) = datagramChannelAndClose
+          _ <- datagramChannel
+                .connect(to)
+                .mapError(ExceptionWrapper)
+                .timeoutFail(ConnectionTimeout(to, connectionTimeout))(connectionTimeout)
+                .toManaged_
+          _ <- logger.info("transport connected to " + to).toManaged_
+        } yield new DatagramChannelOut(datagramChannel, to, requestTimeout, chunkSize, close, self))
+          .provide(self)
+
+      override def bind(
+        addr: SocketAddress
+      )(connectionHandler: ChannelOut => zio.UIO[Unit]): ZManaged[Any, TransportError, ChannelIn] =
+        // There is no way to use the datagram channel as a `ChannelOut` because
+        // it is not connected to any remote address (read on `ChannelOut` cannot be called).
+        // Also, UDP datagram channels are used with `receive`, instead of `read`
+        // because one does not need to necessarily care where the data comes from.
+        // Calling `read` on a connected datagram channel only allows receiving bytes from
+        // the remote address.
+        DatagramChannel()
+          .flatMap(dc => dc.bind(addr).toManaged_.as(dc))
+          .mapError(BindFailed(addr, _))
+          .withEarlyRelease
+          .onExit { _ =>
+            logger.info("shutting down udp channel")
+          }
+          .map {
+            case (close, datagramChannel) =>
+              new DatagramChannelIn(datagramChannel, close)
+          }
+    }
+  }
+
+  private class DatagramChannelOut(
+    datagramChannel: DatagramChannel,
+    remoteAddress: SocketAddress,
+    requestTimeout: Duration,
+    chunkSize: Int,
+    finalizer: URIO[Any, Any],
+    clock: Clock
+  ) extends ChannelOut {
+
+    override def isOpen: ZIO[Any, TransportError, Boolean] =
+      datagramChannel.isOpen
+
+    override def read: ZIO[Any, TransportError, Chunk[Byte]] =
```

Comment:
Save some typing, `IO[TransportError, Chunk[Byte]]`. Also, it can be `val` (the same applies for all other effect defs).